### PR TITLE
Add a list of downstream repos, and script to update

### PR DESCRIPTION
**Pull request by MichaelChirico**
Related: #1292

### CODE REVIEW

File: revdep/get_urls.R
Comment (AshesITR) [inline revdep/get_urls.R:7]: `\blintr\b` is a more compact regex for this.

File: revdep/get_urls.R
Comment (AshesITR) [inline revdep/get_urls.R:33]: `read.dcf(..., fields = "URL")` from the raw description (`/DESCRIPTION` instead of `/index.html`) seems more elegant.
NB requires passing `file("https://...", open = "r")` to `read.dcf` because it doesn't automatically support URLs.

File: revdep/reverse-imports-urls
Comment (AshesITR) [inline revdep/reverse-imports-urls:1]: This file should be `.gitignore`'d imo, compare `revdepcheck` which only versions `failures.md`, `problems.md` and `README.md`.

File: revdep/get_urls.R
Comment (AshesITR) [inline revdep/get_urls.R:1]: I'd put this in `.dev` - `revdep` is used by `revdepcheck`. We shouldn't interfere with its folder structure IMO.

Comment (MichaelChirico) [reply revdep/get_urls.R:1]: oh, that was my first thought, but then I saw this folder. didn't realize it's related to `revdepcheck`

File: revdep/reverse-imports-urls
Comment (MichaelChirico) [reply revdep/reverse-imports-urls:1]: that makes sense, I did have in mind a way for users to report their repos to us if they're not linked on CRAN. but that should be in a separate file since this one's automatic and the other manual. maybe reverse-import-urls-(public vs private) ?

File: revdep/get_urls.R
Comment (MichaelChirico) [reply revdep/get_urls.R]: great! much better

File: revdep/reverse-imports-urls
Comment (AshesITR) [reply revdep/reverse-imports-urls]: Maybe something like `extra-check-urls` could be developed if and when the need arises.

**Comment by AshesITR** [comment]
I noticed `.dev` isn't covered by our `lintr` suite. Do we want to do something about it?

File: .dev/get_urls.R
Comment (AshesITR) [inline .dev/get_urls.R:5]: Use `<-' for assignments

File: .dev/get_urls.R
Comment (AshesITR) [inline .dev/get_urls.R:45]: Where we could add manual URLs.
Would be nice to automatically get a list of packages with missing git repos.
I noticed `fst` and `fstcore` for example.
Those are at `https://github.com/fstpackage/fst/` and `https://github.com/fstpackage/fstcore` respectively.

Comment (AshesITR) [reply .dev/get_urls.R:45]: Could crawl `BugReports` as well, that would have found fst: `"https://github.com/fstpackage/fst/issues"`

File: .dev/get_urls.R
Comment (AshesITR) [inline .dev/get_urls.R:28]: No need for global matching. You can use `regexpr()`.

File: .dev/get_urls.R
Comment (MichaelChirico) [reply .dev/get_urls.R:45]: we could also back up to the cran org mirrors on GitHub, emailing the maintainers if they don't have a repo listed

Comment (MichaelChirico) [reply .dev/get_urls.R:45]: problem is we won't see their .lintr... we could default to running _all_ our linters for those repos

**Comment by MichaelChirico** [comment]
I guess we should also include a recommendation to add us as Suggests in the README? there's typically CRAN requirement all but it'll make things easier

File: .dev/get_urls.R
Comment (AshesITR) [reply .dev/get_urls.R:45]: Let's see how many packages need to be manually found after crawling BugReports. Appsilon packages, e.g. shiny.semantic also provide a github issue link.

I wouldn't loose sleep if a few downstream packages can only be tested on the default lintr config.

File: revdep/reverse-imports-urls
Comment (MichaelChirico) [reply revdep/reverse-imports-urls]: i'll leave it in for now for easier reviewing but will remove+ignore before merging this

**Comment by MichaelChirico** [comment]
> I noticed .dev isn't covered by our lintr suite. Do we want to do something about it?

i think it's fine... not really production scripts, fine if they're a little rough around the edges

File: .dev/get_urls.R
Comment (MichaelChirico) [reply .dev/get_urls.R:45]: for the bugreports one... probably easier to just put those in the extra- manual file

File: .dev/get_urls.R
Comment (AshesITR) [inline .dev/get_urls.R:24]: If you do `c("URL", "BugReports")` and then `paste(drop(pkg_data), collapse = "\n")`, you'll find a bunch more packages.

Comment (MichaelChirico) [reply .dev/get_urls.R:24]: right but for the next step well also have to post-process the URLs 😩

but ok you convinced me. it's only a few extra lines.

Comment (AshesITR) [reply .dev/get_urls.R:24]: Postprocessing works without modifications because the /issues path isn't part of the match :)

Comment (MichaelChirico) [reply .dev/get_urls.R:24]: true... i guess there's an assumption that if both contain a github URL, they're the same, which is acceptable

Comment (AshesITR) [reply .dev/get_urls.R]: Yep (or, at least, the one listed under URL is the correct one)

File: .dev/reverse-imports-no-repos
Comment (MichaelChirico) [inline .dev/reverse-imports-no-repos:2]: definitely also want to .gitignore this one to avoid posting emails (even thought they're easily available from CRAN) -- again keeping it in the PR for now for viz

File: .dev/reverse-imports-no-repos
Comment (AshesITR) [inline .dev/reverse-imports-no-repos:2]: Found it:
https://github.com/maxconway/adaptalint

Comment (MichaelChirico) [reply .dev/reverse-imports-no-repos:2]: nice!

suggest this as the approach for cases like this:

https://github.com/maxconway/adaptalint/pull/2

Comment (MichaelChirico) [reply .dev/reverse-imports-no-repos:2]: definitely highlights some manual friction that's bound to happen as the boundary between the packages we can check automatically and those we track manually changes over time

Comment (MichaelChirico) [reply .dev/reverse-imports-no-repos:2]: Added more PRs:

 - https://github.com/cloudyr/aws.alexa/pull/18
 - https://github.com/compTAG/r-tda/pull/41
 - https://github.com/openpharma/DataFakeR/pull/3
 - https://github.com/phac-nml/plasmidprofiler/pull/1
 
Two more are already added in the dev versions:

 - https://github.com/nhs-r-community/NHSRplotthedots/blob/e54f3d6ab9ce497d2a625e34ab88dae3481ee5c5/DESCRIPTION#L20
 - https://github.com/pharmaverse/admiral/blob/07189fcbbd13a585456fdd826c5d3a7bfe92fe68/DESCRIPTION#L48

Since there's so few left I'll just e-mail the maintainers of the rest:

 - autoharp
 - biolink
 - ConNEcT
 - datarobot
 - DominoDataCapture
 - SamplerCompare
 - wavefunction

Comment (MichaelChirico) [reply .dev/reverse-imports-no-repos:2]: `SamplerCompare`/`wavefunction`, `autoharp`, and `datarobot` maintainers all wrote back confirming no public repo, so I think it's safe to just use the CRAN mirror as a back-up for those and just report deltas.

**Comment by MichaelChirico** [comment]
Added current output to the PR description. Feel free to chime in with more things you'd like to see/collect.

Still haven't had a chance to read the deltas in detail. But did notice that we are doing ~30% better for overall lint performance (even though edgarWebR is still way slower, I think we can ignore that, esp. if this PR is merged: https://github.com/mwaldstein/edgarWebR/pull/24) :tada:

File: .dev/revdep-repos
Comment (AshesITR) [inline .dev/revdep-repos:2]: nit: https:// everywhere?

File: .dev/revdep_compare_releases.R
Comment (AshesITR) [inline .dev/revdep_compare_releases.R:167]: this is a bit verbose.
Maybe also reorder to `linter, line, package, filename`?
The actual line of code seems more important to assess accuracy.

File: .dev/revdep-repos
Comment (MichaelChirico) [reply .dev/revdep-repos:2]: this is taken directly from CRAN, but I have filed PRs to do this, so it'll be picked up eventually:

https://github.com/soodoku/abbyyR/pull/13
https://github.com/soodoku/tuber/pull/119

File: .dev/revdep_compare_releases.R
Comment (MichaelChirico) [reply .dev/revdep_compare_releases.R:167]: this is mainly for our own inspection, locally, so I'm not so worried about verbosity, though the column ordering can be improved.

the next step is to produce e-mails with the delta for each individual package -- there we'll be more careful about presentation.

Comment (MichaelChirico) [reply .dev/revdep_compare_releases.R]: column order didn't matter too much -- the issue is there are v. wide lines that always bump the other columns. so I combined `line [linter]` as one column to force them always to appear together (ditto for `package:filename`)

**Comment by MichaelChirico** [comment]
@AshesITR PTAL. Added the e-mail templates -- these again will not be stored on GH (though the `revdep-email-template` file will be).

I'll put out a call for new repos now (I prefer if downstreams declare themselves to us, btw, rather than getting in downstreams' hair uninvited and/or dealing with low-quality/unmaintained repos [it's also a bit more of a security risk]).

We can fine-tune the output a bit (or handle any bigger feedback), then I'll remove the temp artefacts & we can merge

**Comment by AshesITR** [comment]
I took a glimpse now.

* Some maintainers might get multiple mails. Do we want to bundle all info in one mail?
* The lint comparisons for `brace_linter()` and its predecessors always show up as a change, but I don't think this is material. Can we somehow exclude these sporadic hits?
* Would it be worthwhile to dynamically list what is attached? e.g. if a failure is detected.
* For the failures pertaining to a removed deprecated function, we could repeat the new linter to ease transition after an update to 3.0.0 has already been made. I see `absolute_paths_linter()` causing a few failures.

Also, maybe a good idea to run the checks with a soft-deprecated shim for the removed linters so we can check the remainder of the packages using now removed linters.
IINM we can achieve this by supplying the removed linter functions in globalenv() before calling lint_package().

**Comment by MichaelChirico** [comment]
> Some maintainers might get multiple mails. Do we want to bundle all info in one mail?

Do you know what `revdepcheck` does here? i can see benefits of both ways. currently there are at most a few e-mails to be received so it may not be worth over-engineering...

> The lint comparisons for brace_linter() and its predecessors always show up as a change, but I don't think this is material. Can we somehow exclude these sporadic hits?

Note that the differences are not matching on linter name, just on package/filename/line_number. So if `brace_linter()` is showing up, either the source marker has moved, or it's a new/removed lint. In any case, it should be surfaced to make sure it's intentional.

> Would it be worthwhile to dynamically list what is attached? e.g. if a failure is detected.

Not following what this would look like...

> For the failures pertaining to a removed deprecated function, we could repeat the new linter to ease transition after an update to 3.0.0 has already been made. I see absolute_paths_linter() causing a few failures.

I think there's been transition enough -- as noted in the NEWS, they've been marked as deprecated already since 2017. Time to pull the plug.

> Also, maybe a good idea to run the checks with a soft-deprecated shim for the removed linters so we can check the remainder of the packages using now removed linters. IINM we can achieve this by supplying the removed linter functions in globalenv() before calling lint_package().

great idea :+1:

**Comment by MichaelChirico** [comment]
OK @AshesITR cleaned up the PR here, should be ready for merge if there are no blockers

File: .dev/revdep-email-template
Comment (AshesITR) [inline .dev/revdep-email-template:8]: "position" feels awkward to me. Is "role" appropriate here?

File: .dev/revdep-email-template
Comment (AshesITR) [inline .dev/revdep-email-template:4]: Maybe add a link to `https://lintr.r-lib.org/news/index.html` here?

File: .dev/revdep-email-template
Comment (AshesITR) [inline .dev/revdep-email-template:14]: "We are also attaching..."

**Review by AshesITR — COMMENTED** [review]
Looks good so far, thanks!

File: .dev/revdep_compare_releases.R
Comment (MichaelChirico) [reply .dev/revdep_compare_releases.R:9]: not so magical in this case :)

didn't work on my version of data.table at least

Comment (AshesITR) [reply .dev/revdep_compare_releases.R:9]: Then `skip = "package,repo"` would be more expressive I think.

**Review by AshesITR — APPROVED** [review]

**Pull request merged**
